src/org/mozilla/javascript/GetterSlot.java

```diff
@@ -0,0 +1,194 @@
+        String fName = name == null ? "f" : name.toString();
+        if (getter != null) {
+            Function f = getter.asGetterFunction(fName, scope);
+            desc.defineProperty("get", f == null ? Undefined.instance : f, ScriptableObject.EMPTY);
+        }
+        if (setter != null) {
+            Function f = setter.asSetterFunction(fName, scope);
+            desc.defineProperty("set", f == null ? Undefined.instance : f, ScriptableObject.EMPTY);
+        }
+        return desc;
+    }
+
+    @Override
+    public boolean setValue(Object value, Scriptable owner, Scriptable start) {
+        if (setter == null) {
+            if (getter != null) {
+                throwNoSetterException(start, value);
+                return true;
+            }
+        } else {
+            return setter.setValue(value, owner, start);
+        }
+        return super.setValue(value, owner, start);
+    }
+
+    @Override
+    public Object getValue(Scriptable start) {
+        if (getter != null) {
+            return getter.getValue(start);
+        }
+        return super.getValue(start);
+    }
+
+    @Override
+    Function getSetterFunction(String name, Scriptable scope) {
+        if (setter == null) {
+            return null;
+        }
+        return setter.asSetterFunction(name, scope);
+    }
+
+    @Override
+    Function getGetterFunction(String name, Scriptable scope) {
+        if (getter == null) {
+            return null;
+        }
+        return getter.asGetterFunction(name, scope);
+    }
+
+    abstract static class Getter {
+        abstract Object getValue(Scriptable start);
+
+        abstract Function asGetterFunction(final String name, final Scriptable scope);
+    }
+
+    /** This is a Getter that delegates to a Java function via a MemberBox. */
+    static final class MemberBoxGetter extends Getter {
+        final MemberBox member;
+
+        MemberBoxGetter(MemberBox member) {
+            this.member = member;
+        }
+
+        Object getValue(Scriptable start) {
+            if (member.delegateTo == null) {
+                return member.invoke(start, ScriptRuntime.emptyArgs);
+            }
+            return member.invoke(member.delegateTo, new Object[] {start});
+        }
+
+        Function asGetterFunction(String name, Scriptable scope) {
+            return member.asGetterFunction(name, scope);
+        }
+    }
+
+    /** This is a getter that delegates to a JavaScript function. */
+    static final class FunctionGetter extends Getter {
+        // The value of the function might actually be Undefined, so we need an Object here.
+        final Object target;
+
+        FunctionGetter(Object target) {
+            this.target = target;
+        }
+
+        Object getValue(Scriptable start) {
+            if (target instanceof Function) {
+                Function t = (Function) target;
+                Context cx = Context.getContext();
```

Comment:
Any chance to pass the context from the call tree?

Reply:
Sadly this gets called from Scriptable.get, which has no Context object, and it's been this way for a very long time. I think that changing this now would be pretty huge.